### PR TITLE
Upgrade dependencies 2024-01-22 (#5876, #5906)

### LINKED ISSUE #5876: Upgrade dependencies 2024-01-22
- [x] Update [PyCharm image](https://github.com/DataBiosphere/azul-docker-pycharm)
  - [x] Bump [base image](https://hub.docker.com/_/debian/tags?name=bullseye) tag (only same Debian release), if possible
  - [x] Bump upstream version, if possible
  - [x] Bump internal version
  - [x] Build and test new image locally with Azul's `make format`
  - [x] Remove unused dependencies with high or critical CVEs
  - [x] Push commit to GitHub (directly to master branch, no PR needed)
  - [x] GH Action workflow succeeded
  - [x] Image is available on [DockerHub](https://hub.docker.com/repository/docker/ucscgi/azul-pycharm) 
- [x] Update [Elasticsearch image](https://github.com/DataBiosphere/azul-docker-elasticsearch)
  - [x] Bump [base image](https://hub.docker.com/_/elasticsearch/tags) tag (only minor and patch versions), if possible
  - [x] Bump internal version 
  - [x] Build and test new image locally with Azul's `make test`
  - [x] Remove unused dependencies with high or critical CVEs
  - [x] Push commit to GitHub (directly to master branch, no PR needed)
  - [x] GH Action workflow succeeded
  - [x] Image is available on [DockerHub](https://hub.docker.com/repository/docker/ucscgi/azul-elasticsearch) 
- [x] Create Azul PR, connected to this issue, with … 
  - [x] … changes to `requirements*.txt` from open Dependabot PRs, one commit per PR
  - [x] … update to [Python](https://hub.docker.com/_/python/tags) (only patch versions) <sub>or no update available</sub>
  - [x] … update to [Terraform](https://hub.docker.com/r/hashicorp/terraform/tags) (only patch versions) <sub>or no update available</sub>
  - [x] … new [PyCharm image](https://hub.docker.com/repository/docker/ucscgi/azul-pycharm)
  - [x] … new [Elasticsearch image](https://hub.docker.com/repository/docker/ucscgi/azul-elasticsearch)
  - [x] … update to [Docker images](https://hub.docker.com/_/docker/tags) (only minor and patch versions) <sub>or no update available</sub>
  - [x] … update to [GitLab](https://hub.docker.com/r/gitlab/gitlab-ce/tags) & [GitLab runner images](https://hub.docker.com/r/gitlab/gitlab-runner/tags) <sub>or no update available</sub>
  - [x] … update to [ClamAV image](https://hub.docker.com/r/clamav/clamav/tags) <sub>or no update available</sub>
  - [x] … update to [GitLab AMI](https://github.com/DataBiosphere/azul/blob/develop/OPERATOR.rst#updating-the-ami-for-gitlab-instances) <sub>or no update available</sub>
- [x] Delete obsolete image tags from DockerHub (but consider that `prod` may not use the latest image) … 
  - [x] … for [PyCharm image](https://hub.docker.com/repository/docker/ucscgi/azul-pycharm)
  - [x] … for [Elasticsearch image](https://hub.docker.com/repository/docker/ucscgi/azul-elasticsearch)
- [x] Created tickets for any deferred updates to …
  - [x] … to next major or minor Python version <sub>or such ticket already exists</sub> (#5736)
  - [x] … to next major Docker version <sub>or such ticket already exists</sub> (#5865)
  - [x] … to next major or minor Terraform version <sub>or such ticket already exists</sub> (#5893)

### PR DESCRIPTION
**Pull request by dsotirho-ucsc**
Connected issue: #5876, #5906

### Author

- [x] Target branch is `develop`
- [x] Name of PR branch matches `upgrades/yyyy-mm-dd`
- [x] On ZenHub, PR is connected to the upgrade issue it resolves
- [x] PR title matches `Upgrade dependencies yyyy-mm-dd`
- [x] PR title references the connected issue

### Author (upgrading deployments)

- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading deployments</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `no demo`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to *Approved* column
- [x] PR is assigned to current operator

### Operator (before pushing merge the commit)

- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Selected `dev.shared` and ran `CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR does not change any Docker image versions</sub>
- [x] Selected `dev.gitlab` and ran `CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR does not include any changes to files in terraform/gitlab</sub>
- [x] Selected `anvildev.shared` and ran `CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR does not change any Docker image versions</sub>
- [x] Selected `anvildev.gitlab` and ran `CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR does not include any changes to files in terraform/gitlab</sub>
- [x] Selected `anvilprod.shared` and ran `CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR does not change any Docker image versions</sub>
- [x] Selected `anvilprod.gitlab` and ran `CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR does not include any changes to files in terraform/gitlab</sub>
- [x] Assigned system administrator <sub>or this PR does not include any changes to files in terraform/gitlab</sub>
- [x] Checked the items in the next section <sub>or this PR includes changes to files in terraform/gitlab</sub>

### System administrator

- [x] Background migrations for `dev.gitlab` are complete <sub>or this PR does not include any changes to files in terraform/gitlab</sub>
- [x] Background migrations for `anvildev.gitlab` are complete <sub>or this PR does not include any changes to files in terraform/gitlab</sub>
- [x] Background migrations for `anvilprod.gitlab` are complete <sub>or this PR does not include any changes to files in terraform/gitlab</sub>
- [x] PR is assigned to operator

### Operator (before pushing merge the commit)

- [x] Selected `dev.gitlab` and ran `make -C terraform/gitlab/runner` <sub>or this PR does not change `azul_docker_version`</sub>
- [x] Selected `anvildev.gitlab` and ran `make -C terraform/gitlab/runner` <sub>or this PR does not change `azul_docker_version`</sub>
- [x] Selected `anvilprod.gitlab` and ran `make -C terraform/gitlab/runner` <sub>or this PR does not change `azul_docker_version`</sub>
- [x] Added `sandbox` label
- [x] Pushed PR branch to GitLab `dev`
- [x] Pushed PR branch to GitLab `anvildev`
- [x] Pushed PR branch to GitLab `anvilprod`
- [x] Build passes in `sandbox` deployment
- [x] Build passes in `anvilbox` deployment
- [x] Build passes in `hammerbox` deployment
- [x] Reviewed build logs for anomalies in `sandbox` deployment
- [x] Reviewed build logs for anomalies in `anvilbox` deployment
- [x] Reviewed build logs for anomalies in `hammerbox` deployment
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but exclude any `p` tags</sub>
- [x] Moved connected issue to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Pushed merge commit to GitLab `anvilprod`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [x] Build passes on GitLab `anvildev`
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Build passes on GitLab `anvilprod`
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`
- [x] Selected `dev.shared` and ran `make -C terraform/shared apply` <sub>or this PR does not change any Docker image versions</sub>
- [x] Selected `anvildev.shared` and ran `make -C terraform/shared apply` <sub>or this PR does not change any Docker image versions</sub>
- [x] Selected `anvilprod.shared` and ran `make -C terraform/shared apply` <sub>or this PR does not change any Docker image versions</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

### Operator

- [x] Ran `script/export_inspector_findings.py` against `anvilprod`, imported results to [Google Sheet](https://docs.google.com/spreadsheets/d/1RWF7g5wRKWPGovLw4jpJGX_XMi8aWLXLOvvE5rxqgH8) and posted screenshot of relevant<sup>1</sup> findings as a comment on the connected issue.
- [x] PR is assigned to system administrator

<sup>1</sup>A relevant finding is a high or critical vulnerability in an image
that is used within the security boundary. Images not used within the boundary
are tracked in `azul.docker_images` under a key starting with `_`.

### System administrator

- [x] No currently reported vulnerability requires immediate attention
- [x] PR is assigned to no one

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by dsotirho-ucsc** [comment]
Run a test:
- disable docker
- deploy & run integration_test in personal deployment
- if successful, move `shared (apply_keep_unused)` checkboxes just before deploy `gitlab` checkboxes & file a ticket to update PR template.

**Comment by dsotirho-ucsc** [comment]
> Run a test:
> 
> * disable docker
> * deploy & run integration_test in personal deployment
> * if successful, move `shared (apply_keep_unused)` checkboxes just before deploy `gitlab` checkboxes & file a ticket to update PR template.

Test was successful.
```
(.venv) daniel@Crispin ~/repo/azul2 $ git s
On branch upgrades/2024-01-22
Your branch is up to date with 'github/upgrades/2024-01-22'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.terraform-version

nothing added to commit but untracked files present (use "git add" to track)
(.venv) daniel@Crispin ~/repo/azul2 $
(.venv) daniel@Crispin ~/repo/azul2 $ make deploy
/Library/Developer/CommandLineTools/usr/bin/make -C lambdas
/Library/Developer/CommandLineTools/usr/bin/make -C layer layer
python /Users/daniel/repo/azul2/scripts/check_branch.py
python .chalice/config.json.template.py .chalice/config.json
Creating .chalice/config.json
python /Users/daniel/repo/azul2/scripts/stage_layer.py
2024-01-29 10:09:41,042    INFO MainThread botocore.credentials: Found credentials in shared credentials file: ~/.aws/credentials
2024-01-29 10:09:41,583    INFO MainThread azul.lambda_layer: Checking for dependencies layer package at s3://edu-ucsc-gi-platform-hca-dev-shared.us-east-1/azul/daniel/lambda_layers/e43d09c184c6d18ddb69a0eaf56a9986846c406f.zip.
…
Plan: 8 to add, 11 to change, 39 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_api_gateway_base_path_mapping.service_1: Destroying... [id=drs.daniel.dev.singlecell.gi.ucsc.edu/]
aws_api_gateway_method_settings.service_0: Destroying... [id=071n66awdf-daniel-*/*]
aws_cloudwatch_metric_alarm.service_throttles: Destroying... [id=azul-service_throttles-daniel.alarm]
…
Apply complete! Resources: 8 added, 11 changed, 39 destroyed.
python /Users/daniel/repo/azul2/scripts/post_deploy_tdr.py
2024-01-29 10:16:22,364    INFO MainThread botocore.credentials: Found credentials in shared credentials file: ~/.aws/credentials
2024-01-29 10:16:23,038    INFO MainThread azul.terra: Making POST request to 'https://sam.dsde-dev.broadinstitute.org/register/user/v1'
2024-01-29 10:16:23,038   DEBUG MainThread azul.terra: … with request body ''
…

(.venv) daniel@Crispin ~/repo/azul2 $ 
(.venv) daniel@Crispin ~/repo/azul2 $ docker ps
Cannot connect to the Docker daemon at unix:///Users/daniel/.docker/run/docker.sock. Is the docker daemon running?
(.venv) daniel@Crispin ~/repo/azul2 $ 
(.venv) daniel@Crispin ~/repo/azul2 $ make integration_test
python /Users/daniel/repo/azul2/scripts/check_branch.py
python /Users/daniel/repo/azul2/lambdas/service/.chalice/config.json.template.py /Users/daniel/repo/azul2/lambdas/service/.chalice/config.json
[INFO]	2024-01-29T10:27:54.458Z	00010ca1-b0ba-466f-8c58-dabbad000000	botocore.credentials	Found credentials in shared credentials file: ~/.aws/credentials
Creating /Users/daniel/repo/azul2/lambdas/service/.chalice/config.json
python -m unittest -v integration_test
2024-01-29 10:27:56,616    INFO MainThread botocore.credentials: Found credentials in shared credentials file: ~/.aws/credentials
2024-01-29 10:27:57,178   DEBUG MainThread azul.es: Creating ES client [vpc-azul-index-sandbox-mcwjphhhdivigzrsrdmxm2uude.us-east-1.es.amazonaws.com:443]
2024-01-29 10:27:57,221    INFO MainThread botocore.credentials: Found credentials in environment variables.
…
2024-01-29 10:45:32,979   DEBUG MainThread test.integration_test: … with response body b'{"message":"Forbidden"}'
2024-01-29 10:45:32,979    INFO MainThread test.integration_test: Making GET request to 'https://indexer.daniel.dev.singlecell.gi.ucsc.edu/static/..%2Fdoes-not-exist'
2024-01-29 10:45:32,979   DEBUG MainThread test.integration_test: … without request body
2024-01-29 10:45:33,184    INFO MainThread test.integration_test: Got 403 response after 0.204s from GET to https://indexer.daniel.dev.singlecell.gi.ucsc.edu/static/..%2Fdoes-not-exist
2024-01-29 10:45:33,184   DEBUG MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '23', 'Connection': 'keep-alive', 'Date': 'Mon, 29 Jan 2024 18:45:33 GMT', 'x-amzn-RequestId': '90be3a7f-6742-40fb-b8c8-ad880b61166b', 'x-amzn-ErrorType': 'ForbiddenException', 'x-amz-apigw-id': 'SULMHGfVoAMEclQ=', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 e61b74b41588d9216f1bb35848394554.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'SFO20-C1', 'X-Amz-Cf-Id': 'MfNQFHfw2hXRkOotBAPcHVumo2apx8eEGyAfOFsxUmjJggNioLfaQA=='})
2024-01-29 10:45:33,184   DEBUG MainThread test.integration_test: … with response body b'{"message":"Forbidden"}'
ok

----------------------------------------------------------------------
Ran 15 tests in 1056.592s

OK (skipped=3)
(.venv) daniel@Crispin ~/repo/azul2 $
```

**Comment by dsotirho-ucsc** [comment]
[5897_IT_2024-01-29.txt](https://github.com/DataBiosphere/azul/files/14090388/5897_IT_2024-01-29.txt)

**Comment by dsotirho-ucsc** [comment]
[5897_integration-test_2024-01-29.txt](https://github.com/DataBiosphere/azul/files/14090834/5897_integration-test_2024-01-29.txt)

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
There's an unchecked item in the CL in the issue.

**Comment by dsotirho-ucsc** [comment]
[5897_integration-test_2024-01-29.txt](https://github.com/DataBiosphere/azul/files/14091526/5897_integration-test_2024-01-29.txt)

File: UPGRADING.rst [hunk 1]
```diff
@@ -40,6 +40,11 @@ lower and stable deployments, perform the second phase using the ``apply``
 Makefile target after the merge commit was successfully built on the GitLab
 instance in that deployment.
 
+Deploy the ``gitlab`` component of any main deployment just after pushing the PR
+branch to GitHub. Run ``make -C terraform/gitlab/runner`` just before pushing
+the merge commit to the GitLab instance in that deployment.
```
Comment (hannes-ucsc) [inline UPGRADING.rst:45]: The second sentence is in the wrong commit.

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
The previous review should have requested changes.

**Comment by hannes-ucsc** [comment]
Please set tfenv version with environment variables so you can have a different version per deployment and don't have a working copy that's permanently dirtied by a file called `.terraform-version`.

**Review by hannes-ucsc — DISMISSED** [review]
I think I can live with the wrong commit structure.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by dsotirho-ucsc** [comment]
https://docs.google.com/spreadsheets/d/1RWF7g5wRKWPGovLw4jpJGX_XMi8aWLXLOvvE5rxqgH8/edit#gid=1432851400

image

**Comment by hannes-ucsc** [comment]
CVE-2018-12976: From https://groups.google.com/g/golang-announce/c/4rpTbfzYB1k/m/no6MEwlQAwAJ

> [godoc.org](http://godoc.org/) is patched, and this only affects you if you run your own instance of gddo.

While the vulnerable version of the `gddo` module is included in the `dockerd` binary (it is [vendored](https://github.com/moby/moby/blob/master/vendor.mod#L47)), and dockerd is running as a daemon, it does not represent an instance of `gddo`, and is therefore not vulnerable, according to the above announcement.

**Comment by hannes-ucsc** [comment]
Please post screenshot to issue, not PR. See CL item.

**Pull request merged**